### PR TITLE
ScriptEngine.eval() forwards Error instead of new ScriptException

### PR DESCRIPTION
**Pull request by rjolly**

### CODE REVIEW

File: src/repl/scala/tools/nsc/interpreter/IMain.scala [hunk 1]
```diff
@@ -597,11 +597,19 @@ class IMain(@BeanProperty val factory: ScriptEngineFactory, initialSettings: Set
   private class WrappedRequest(val req: Request) extends CompiledScript {
     var recorded = false
 
+    /** In Java we would have to wrap any checked exception in the declared
+     *  ScriptException. Runtime exceptions and errors would be ok and would
+     *  not need to be catched. So let us do the same in Scala : catch and
```
Comment (retronym) [inline src/repl/scala/tools/nsc/interpreter/IMain.scala]: caught

File: src/repl/scala/tools/nsc/interpreter/IMain.scala [hunk 2]
```diff
@@ -597,11 +597,19 @@ class IMain(@BeanProperty val factory: ScriptEngineFactory, initialSettings: Set
...
+     *  escape. We could have wrapped runtime exceptions just like other
+     *  exceptions in ScriptException, this is a choice.
+     */
     @throws(classOf[ScriptException])
     def eval(context: ScriptContext): Object = {
       val result = req.lineRep.evalEither match {
+        case Left(e: RuntimeException) => throw e
         case Left(e: Exception) => throw new ScriptException(e)
-        case Left(_) => throw new ScriptException("run-time error")
+        case Left(e) => throw e
```
Comment (retronym) [inline src/repl/scala/tools/nsc/interpreter/IMain.scala:612]: This calls for a utility method, you could add it here: `src/compiler/scala/tools/nsc/util/package.scala`.

``` scala
def wrapUncheckedException(t: Throwable, wrap: Throwable => Throwable = new RuntimeException(_)): Throwable = t match {
  case re: RuntimeException => re
  case e: Exception => wrap(e)
  case t => t
}
```

**Comment by rjolly** [comment]
This is very scripting specific (scripts need clear error reporting) but I am not sure it is useful elsewhere. We are starting to fix Scala fondational choices regarding exceptions here...

**Comment by adriaanm** [comment]
LGTM, though Jason's refactoring suggestion stands for future refinement

**Pull request merged**